bootstore/src/schemes/v0/request_manager.rs

```diff
@@ -0,0 +1,401 @@
+                    request_id,
+                    self.requests.remove(&request_id).unwrap(),
+                );
+            } else {
+                // Put the last request back. We are done.
+                self.expiry_to_id.insert(expiry, request_id);
+                break;
+            }
+        }
+        expired
+    }
+
+    /// Return `Some(true)` if initialization completed, `Some(false) if it did not
+    ///
+    /// Return `None` if there is no active rack initialization ongoing
+    pub fn on_init_ack(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+    ) -> Option<bool> {
+        if let Some(TrackableRequest::InitRack { acks, .. }) =
+            self.requests.get_mut(&request_id)
+        {
+            acks.received.insert(from);
+            if acks.received == acks.expected {
+                self.expiry_to_id.retain(|_, id| *id != request_id);
+                self.requests.remove(&request_id);
+                Some(true)
+            } else {
+                Some(false)
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Return the `Some(request)` if a threshold of acks has been received.
+    /// Otherwise return `None`
+    pub fn on_share(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+        share: Share,
+    ) -> Option<TrackableRequest> {
+        let acks = match self.requests.get_mut(&request_id) {
+            Some(TrackableRequest::LoadRackSecret { acks, .. }) => acks,
+            Some(TrackableRequest::LearnReceived { acks, .. }) => acks,
+            _ => return None,
+        };
+
+        acks.received.insert(from, share);
+        // We already have our own share to be used to reconstruct the secret
+        if acks.received.len() == (acks.threshold - 1) as usize {
+            self.expiry_to_id.retain(|_, id| *id != request_id);
+            self.requests.remove(&request_id)
+        } else {
+            None
+        }
+    }
+
+    /// Return true if there is a `LearnSent` for the given `request_id`, false
+    /// otherwise.
+    pub fn on_pkg(&mut self, request_id: Uuid) -> bool {
+        if let Some(TrackableRequest::LearnSent { .. }) =
+            self.requests.get_mut(&request_id)
+        {
+            self.expiry_to_id.retain(|_, id| *id != request_id);
+            self.requests.remove(&request_id);
+            true
+        } else {
+            false
+        }
+    }
+
+    /// If there are outstanding requests and this peer has not acknowledged
+    /// the given request then send the request to the peer.
+    pub fn on_connected(&mut self, peer_id: &Baseboard) {
+        for (request_id, request) in &self.requests {
+            match request {
+                TrackableRequest::InitRack { packages, acks, .. } => {
+                    if acks.received.contains(peer_id) {
+                        continue;
+                    }
+                    if let Some(pkg) = packages.get(peer_id) {
```

Comment:
Do we need to add `peer_id` to `acks.expected` here?

Reply:
No, but this along with your [prior comment](https://github.com/oxidecomputer/omicron/pull/3385#discussion_r1256546031) is a **great** catch. If there is a node that sends us an ack, for example because it was a learner that connected and wasn't an explicit initial member but we still sent it a request here, we would end up in the situation above, where we get permanently stuck.  

What should actually be done here is to not store any acks from peers that are not members `expected` and also not send those peers `Init` requests.

Reply:
Fixed in [774ac90](https://github.com/oxidecomputer/omicron/pull/3385/commits/774ac90417f756fdbf8d141532b013f716cb2e43)